Simplenote/NotesListController.swift

```diff
@@ -0,0 +1,213 @@
+                                                               sortedBy: state.descriptorsForNotes(sortMode: sortMode))
+
+    /// FSM: Active State
+    ///
+    private(set) var state: NotesListState = .results {
+        didSet {
+            guard oldValue != state else {
+                return
+            }
+
+            refreshEverything()
+        }
+    }
+
+    /// Filter: Applied when we're not in Search Mode
+    ///
+    var filter: NotesListFilter = .everything {
+        didSet {
+            guard oldValue != filter else {
+                return
+            }
+
+            refreshPredicates()
+        }
+    }
+
+    /// Refreshes the Internal State, so that Search Results matching the specified Keyword will be filtered out.
+    /// - Important: It's up to the caller to invoke `performFetch`!!
+    ///
+    var searchKeyword: String? {
+        get {
+            guard case let .searching(keyword) = self.state else {
+                return nil
+            }
+
+            return keyword
+        }
+        set {
+            guard let keyword = newValue, !keyword.isEmpty else {
+                state = .results
+                return
+            }
+
+            state = .searching(keyword: keyword)
+        }
+    }
+
+    /// SortMode: Results Mode
+    ///
+    var sortMode: SortMode = .alphabeticallyAscending {
+        didSet {
+            guard oldValue != sortMode else {
+                return
+            }
+
+            refreshSortDescriptors()
+        }
+    }
+
+    /// SortMode: Search Mode
+    ///
+    var searchSortMode: SortMode = .alphabeticallyAscending {
+        didSet {
+            guard case .searching = state, oldValue != searchSortMode else {
+                return
+            }
+
+            refreshSortDescriptors()
+        }
+    }
+
+    /// Relays back willChangeContent Events
+    ///
+    var onWillChangeContent: (() -> Void)?
+
+    /// Relays back didChangeContent Events
+    ///
+    var onDidChangeContent: ((_ rowsChangeset: ResultsObjectsChangeset) -> Void)?
+
+
+    /// Designated Initializer
+    ///
+    init(viewContext: NSManagedObjectContext) {
+        self.viewContext = viewContext
+        super.init()
+        startListeningToNoteEvents()
+    }
+}
+
+
+// MARK: - Public API
+//
+extension NotesListController {
+
+    /// Number of the notes we've got!
+    ///
+    @objc
+    var numberOfNotes: Int {
+        notesController.numberOfObjects
+    }
+
+    /// Returns all of the Retrieved Notes
+    ///
+    var retrievedNotes: [Note] {
+        notesController.fetchedObjects
+    }
+
+    /// Returns the index of a given note (if any)
+    ///
+    func indexOfNote(withSimperiumKey key: String) -> Int? {
+        for (index, note) in notesController.fetchedObjects.enumerated() where note.simperiumKey == key {
+            return index
+        }
+
+        return nil
+    }
```

Comment:
I tend to avoid APIs that use comparator blocks (because, unless Swift optimizes, they should have a performance hit).

Nevertheless, compact wins, updated! Thanks for the suggestion!